test/plugins/iamauth.js

```diff
@@ -0,0 +1,588 @@
+      iamMocks.done();
+      cloudantMocks.done();
+      done();
+    });
+  });
+
+  it('performs request and returns error', function(done) {
+    if (process.env.NOCK_OFF) {
+      this.skip();
+    }
+
+    var iamMocks = nock(TOKEN_SERVER)
+      .post('/oidc/token', {
+        'grant_type': 'urn:ibm:params:oauth:grant-type:apikey',
+        'response_type': 'cloud_iam',
+        'apikey': IAM_API_KEY
+      })
+      .reply(200, MOCK_OIDC_TOKEN_RESPONSE);
+
+    var cloudantMocks = nock(SERVER)
+      .post('/_iam_session', {access_token: MOCK_ACCESS_TOKEN})
+      .reply(200, {ok: true}, MOCK_SET_IAM_SESSION_HEADER)
+      .get(DBNAME)
+      .replyWithError({code: 'ECONNRESET', message: 'socket hang up'});
+
+    var cloudantClient = new Client({ iamApiKey: IAM_API_KEY, plugin: 'iamauth' });
+    var req = { url: SERVER + DBNAME, method: 'GET' };
+    cloudantClient.request(req, function(err, resp, data) {
+      assert.equal(err.code, 'ECONNRESET');
+      assert.equal(err.message, 'socket hang up');
+      iamMocks.done();
+      cloudantMocks.done();
+      done();
+    });
+  });
+
+  it('retries access token post on error and returns 200 response', function(done) {
+    if (process.env.NOCK_OFF) {
+      this.skip();
+    }
+
+    var iamMocks = nock(TOKEN_SERVER)
+      .post('/oidc/token', {
+        'grant_type': 'urn:ibm:params:oauth:grant-type:apikey',
+        'response_type': 'cloud_iam',
+        'apikey': IAM_API_KEY
+      })
+      .replyWithError({code: 'ECONNRESET', message: 'socket hang up'})
+      .post('/oidc/token', {
+        'grant_type': 'urn:ibm:params:oauth:grant-type:apikey',
+        'response_type': 'cloud_iam',
+        'apikey': IAM_API_KEY
+      })
+      .reply(200, MOCK_OIDC_TOKEN_RESPONSE);
+
+    var cloudantMocks = nock(SERVER)
+      .post('/_iam_session', {access_token: MOCK_ACCESS_TOKEN})
+      .reply(200, {ok: true}, MOCK_SET_IAM_SESSION_HEADER)
+      .get(DBNAME)
+      .reply(200, {doc_count: 0});
+
+    var cloudantClient = new Client({ iamApiKey: IAM_API_KEY, plugin: 'iamauth' });
+    var req = { url: SERVER + DBNAME, method: 'GET' };
+    cloudantClient.request(req, function(err, resp, data) {
+      assert.equal(err, null);
+      assert.equal(resp.request.headers.cookie, MOCK_IAM_SESSION);
+      assert.equal(resp.statusCode, 200);
+      assert.ok(data.indexOf('"doc_count":0') > -1);
+      iamMocks.done();
+      cloudantMocks.done();
+      done();
+    });
+  });
+
+  it('skips IAM authentication if access token returns non-200 response', function(done) {
+    if (process.env.NOCK_OFF) {
+      this.skip();
+    }
+
+    var iamMocks = nock(TOKEN_SERVER)
+      .post('/oidc/token', {
+        'grant_type': 'urn:ibm:params:oauth:grant-type:apikey',
+        'response_type': 'cloud_iam',
+        'apikey': IAM_API_KEY
+      })
+      .reply(500, {error: 'internal_server_error', reason: 'Internal Server Error'});
```

Comment:
I wonder if it is worth changing the body content here to something less CouchDB JSON like. I'm not sure what a 500 would look like from the IAM server, but it's unlikely to follow these patterns. It would provide us some reassurance that we aren't accidentally relying on some JSON processing that uses the known Couch fields when we are exercising this non-CouchDB path.

Reply:
Switched to plaintext error message in bf75551.